File: diameter-of-binary-tree.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
int maxDepth(TreeNode* root) {
  if(!root)
    return 0;
  int depth_left = 0;
  if(root->left)
    depth_left = maxDepth(root->left);
  int depth_right = 0;
  if(root->right)
    depth_right = maxDepth(root->right);
  return 1+max(depth_right, depth_left);
}

int diameterOfBinaryTree2(TreeNode* root) {
  if(!root)
    return 0;
  int case1 = diameterOfBinaryTree2(root->right);
  int case2 = diameterOfBinaryTree2(root->left);
  int case3 = 1 + maxDepth(root->left) + maxDepth(root->right);
  return max(max(case1, case2), case3);
}
int diameterOfBinaryTree(TreeNode* root) {
  return diameterOfBinaryTree2(root)-1;
}
};

File: diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
  def maxDepth(self, root):
    """
    :type root: TreeNode
    :rtype: int
    """
    if not root:
      return 0
    return 1+max(self.maxDepth(root.left), self.maxDepth(root.right))

  def diameterOfBinaryTree2(self, root):
    if not root:
      return 0
    return max(self.diameterOfBinaryTree2(root.left), self.diameterOfBinaryTree2(root.right), 1+self.maxDepth(root.left)+self.maxDepth(root.right))
  
  def diameterOfBinaryTree(self, root):
    if not root:
      return 0
    return self.diameterOfBinaryTree2(root)-1
